nodes.py

```diff
@@ -69,25 +69,87 @@ def check_semantics(self, variables_dict):
             )
             return (1, "")
-        if left_type == Types.String or right_type == Types.String:
+        if right_type in [Types.Bool, Types.String]:
             print(
-                f"ERROR: {operation_name} string type is not allowed (line: {self.line_no}) "
+                f"ERROR: {right_type.value} on the right side of {operation_name} is not allowed (line: {self.line_no})"
             )
             return (1, "")
         if left_type == right_type == Types.Int:
             return (0, Types.Int)
         return (0, Types.Float)
 
+    def __handle_logical_operator(self, operation_name, left_type, right_type):
+        if left_type in [Types.Float, Types.Int, Types.String]:
+            print(
+                f"ERROR: {left_type.value} on the left side of {operation_name} is not allowed (line: {self.line_no})"
+            )
+            return (1, "")
+        if right_type in [Types.Float, Types.Int, Types.String]:
+            print(
+                f"ERROR: {right_type.value} on the right side of {operation_name} is not allowed (line: {self.line_no})"
+            )
+            return (1, "")
+        return (0, Types.Bool)
+
     def __str__(self, indent_level=0):
         return super().__str__(indent_level, f"({self.op})")
 
+    def __str__(self, indent_level=0):
+        return super().__str__(indent_level, f"({self.op})")
+
+    def write_code(self, output_lines):
+        if self.op in ["+", "-", "*", "/"]:
+            left_type, left_mem_id, left_val = self.left.write_code(output_lines)
+            right_type, right_mem_id, right_val = self.right.write_code(output_lines)
+
+            if left_type != right_type:
+                return_type = Types.Float
+                if left_type is Types.Int:
+                    if left_val != "":
+                        output_lines.append(
+                            f"  %{ProgramMemory.mem_counter} = sitofp i32 {left_val} to float"
+                        )
+                    else:
+                        output_lines.append(
+                            f"  %{ProgramMemory.mem_counter} = sitofp i32 %{left_mem_id} to float"
+                        )
+                    left_val = ""
+                    left_mem_id = ProgramMemory.mem_counter
+                    ProgramMemory.mem_counter += 1
+                else:
+                    if right_val != "":
+                        output_lines.append(
+                            f"  %{ProgramMemory.mem_counter} = sitofp i32 {right_val} to float"
+                        )
+                    else:
+                        output_lines.append(
+                            f"  %{ProgramMemory.mem_counter} = sitofp i32 %{right_mem_id} to float"
+                        )
+                    right_val = ""
+                    right_mem_id = ProgramMemory.mem_counter
+                    ProgramMemory.mem_counter += 1
+            else:
+                return_type = left_type
+
```

Comment:
Code in lines 116-152 is doubled, probably because of merge. Lines 116-117 are in 113-114 and lines 119-151 are in 153-185 (in case of this changes).

Reply:
fixed